streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -947,127 +665,428 @@ trait ZSink[-R, +E, +A0, -A, +B] { self =>
+          new SinkPure[E, A0, A, B] {
+            type State = self.State
+            val initialPure                  = self.initialPure
+            def stepPure(state: State, a: A) = if (f(a)) self.stepPure(state, a) else state
+            def extractPure(state: State)    = self.extractPure(state)
+            def cont(state: State)           = self.cont(state)
+          }
+
+        case _ =>
+          new ZSink[R, E, A0, A, B] {
+            type State = self.State
+            val initial                  = self.initial
+            def step(state: State, a: A) = if (f(a)) self.step(state, a) else UIO.succeed(state)
+            def extract(state: State)    = self.extract(state)
+            def cont(state: State)       = self.cont(state)
+          }
+      }
+
+    /**
+     * Effectfully filters the inputs fed to this sink.
+     */
+    final def filterM[R1 <: R, E1 >: E](f: A => ZIO[R1, E1, Boolean]): ZSink[R1, E1, A0, A, B] =
+      new ZSink[R1, E1, A0, A, B] {
+        type State = self.State
+        val initial = self.initial
+
+        def step(state: State, a: A) = f(a).flatMap { b =>
+          if (b) self.step(state, a)
+          else UIO.succeed(state)
+        }
+
+        def extract(state: State) = self.extract(state)
+        def cont(state: State)    = self.cont(state)
+      }
+
+    /**
+     * Filters this sink by the specified predicate, dropping all elements for
+     * which the predicate evaluates to true.
+     */
+    final def filterNot(f: A => Boolean): ZSink[R, E, A0, A, B] =
+      filter(a => !f(a))
+
+    /**
+     * Effectfully filters this sink by the specified predicate, dropping all elements for
+     * which the predicate evaluates to true.
+     */
+    final def filterNotM[R1 <: R, E1 >: E](f: A => ZIO[R1, E1, Boolean]): ZSink[R1, E1, A0, A, B] =
+      filterM(a => f(a).map(!_))
+
+    /**
+     * Runs `n` sinks in parallel, where `n` is the number of possible keys
+     * generated by `f`.
+     */
+    final def keyed[K](f: A => K): ZSink[R, E, (K, Chunk[A0]), A, Map[K, B]] =
+      new ZSink[R, E, (K, Chunk[A0]), A, Map[K, B]] {
+        type State = Map[K, self.State]
+
+        val initial =
+          self.initial.map(init => Map.empty[K, self.State].withDefaultValue(init))
+
+        def step(state: State, a: A) = {
+          val k = f(a)
+          self.step(state(k), a).map(s => state + (k -> s))
+        }
+
+        def extract(state: State) =
+          ZIO
+            .foreach(state.toList) {
+              case (k, s) => self.extract(s).map(k -> _)
+            }
+            .map { list =>
+              val results   = list.map { case (k, (b, _)) => (k, b) }.toMap
+              val leftovers = Chunk.fromIterable(list.map { case (k, (_, chunk)) => (k, chunk) }).filter(_._2.notEmpty)
+              (results, leftovers)
+            }
+
+        def cont(state: State) = state.values.forall(self.cont)
+      }
   }
 
   private[ZSink] object internal {
```

Comment:
Can the `Optional` ADT be migrated to `?` as well? OTOH Not sure if `Side` is used in multiple combinators.

Reply:
Yup! Moved it